src/lib/crypto/krb/prng_os.c

```diff
@@ -26,68 +26,49 @@
  */
 
 /*
- * This file implements a PRNG module which relies on the system's /dev/urandom
+ * This file implements a PRNG module which relies on the system's kernel
+ * PRNG, acccessed through getrandom() with fallback to the /dev/urandom
  * device.  An OS packager can select this module given sufficient confidence
  * in the operating system's native PRNG quality.
  */
 
 #include "crypto_int.h"
 
-#define DEVICE "/dev/urandom"
-
-static int fd = -1;
-
 int
 k5_prng_init(void)
 {
-    /* Try to open the random device read-write; if that fails, read-only is
-     * okay. */
-    fd = open(DEVICE, O_RDWR, 0);
-    if (fd == -1)
-        fd = open(DEVICE, O_RDONLY, 0);
-    if (fd == -1)
-        return errno;
     return 0;
 }
 
 void
 k5_prng_cleanup(void)
 {
-    close(fd);
-    fd = -1;
+    return;
 }
 
 krb5_error_code KRB5_CALLCONV
 krb5_c_random_add_entropy(krb5_context context, unsigned int randsource,
                           const krb5_data *indata)
 {
-    krb5_error_code ret;
-
-    ret = krb5int_crypto_init();
-    if (ret)
-        return ret;
-
-    /* Some random devices let user-space processes contribute entropy.  Don't
-     * worry if this fails. */
-    (void)write(fd, indata->data, indata->length);
     return 0;
 }
 
 krb5_error_code KRB5_CALLCONV
 krb5_c_random_make_octets(krb5_context context, krb5_data *outdata)
 {
+    krb5_boolean res;
     char *buf = outdata->data;
     size_t len = outdata->length;
-    ssize_t count;
 
-    while (len > 0) {
-        count = read(fd, buf, len);
-        if (count == 0)         /* Not expected from a random device. */
-            return KRB5_CRYPTO_INTERNAL;
-        if (count == -1)
-            return errno;
-        buf += count;
-        len -= count;
-    }
+    res = k5_get_os_entropy(buf, len, 0);
+
+    if (res == FALSE)
+        return KRB5_CRYPTO_INTERNAL;
+    return 0;
```

Comment:
Use a trinary operator to simplify this.  Also, the blank line here doesn't really separate a code paragraph and should be removed.  Also, buf and len can be removed and their uses expanded out.